File: client/src/game/collision.js
const WALL_HALF_THICKNESS = 0.2; // feet; matches server wallHalfThickness and visual rendering

function pushOutOfSegment(px, py, r, ax, ay, bx, by) {
  const dx = bx - ax, dy = by - ay;
  const lenSq = dx * dx + dy * dy;
  if (lenSq === 0) return [px, py];
  let t = ((px - ax) * dx + (py - ay) * dy) / lenSq;
  if (t < 0) t = 0; else if (t > 1) t = 1;
  const cx = ax + t * dx, cy = ay + t * dy;
  const ex = px - cx, ey = py - cy;
  const dist = Math.sqrt(ex * ex + ey * ey);
  if (dist >= r) return [px, py];
  if (dist === 0) {
    const len = Math.sqrt(lenSq);
    return [px + (-dy / len) * r, py + (dx / len) * r];
  }
  const overlap = r - dist;
  return [px + (ex / dist) * overlap, py + (ey / dist) * overlap];
}

function pushOutOfCircle(px, py, r, cx, cy, barrierR) {
  const dx = px - cx, dy = py - cy;
  const dist = Math.sqrt(dx * dx + dy * dy);
  const minDist = r + barrierR;
  if (dist >= minDist) return [px, py];
  if (dist === 0) return [px, py + minDist];
  const overlap = minDist - dist;
  return [px + (dx / dist) * overlap, py + (dy / dist) * overlap];
}

function cross(ax, ay, bx, by) {
  return ax * by - ay * bx;
}

// Assumes (px,py) is collinear with (ax,ay)-(bx,by); checks it falls within
// the segment's bounding box.
function onSegment(ax, ay, bx, by, px, py) {
  return px >= Math.min(ax, bx) && px <= Math.max(ax, bx) && py >= Math.min(ay, by) && py <= Math.max(ay, by);
}

// Reports whether segments (x1,y1)-(x2,y2) and (x3,y3)-(x4,y4) cross, using
// the standard orientation test. Endpoint-touching counts as crossing.
// Mirrors instanceconfig.segmentsIntersect on the server.
function segmentsIntersect(x1, y1, x2, y2, x3, y3, x4, y4) {
  const d1 = cross(x4 - x3, y4 - y3, x1 - x3, y1 - y3);
  const d2 = cross(x4 - x3, y4 - y3, x2 - x3, y2 - y3);
  const d3 = cross(x2 - x1, y2 - y1, x3 - x1, y3 - y1);
  const d4 = cross(x2 - x1, y2 - y1, x4 - x1, y4 - y1);

  if (((d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0)) && ((d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0))) return true;
  if (d1 === 0 && onSegment(x3, y3, x4, y4, x1, y1)) return true;
  if (d2 === 0 && onSegment(x3, y3, x4, y4, x2, y2)) return true;
  if (d3 === 0 && onSegment(x1, y1, x2, y2, x3, y3)) return true;
  if (d4 === 0 && onSegment(x1, y1, x2, y2, x4, y4)) return true;
  return false;
}

// Reports whether the segment (x1,y1)-(x2,y2) passes within radius of (cx,cy).
// Mirrors instanceconfig.segmentIntersectsCircle on the server.
function segmentIntersectsCircle(x1, y1, x2, y2, cx, cy, radius) {
  const dx = x2 - x1, dy = y2 - y1;
  const lenSq = dx * dx + dy * dy;
  if (lenSq === 0) {
    const ex = cx - x1, ey = cy - y1;
    return ex * ex + ey * ey <= radius * radius;
  }
  let t = ((cx - x1) * dx + (cy - y1) * dy) / lenSq;
  if (t < 0) t = 0; else if (t > 1) t = 1;
  const nearX = x1 + t * dx, nearY = y1 + t * dy;
  const ex = nearX - cx, ey = nearY - cy;
  return ex * ex + ey * ey <= radius * radius;
}

// Reports whether the straight line between (x1,y1) and (x2,y2) is
// unobstructed by any of the given barriers. Mirrors
// instanceconfig.LineOfSightClear on the server.
export function hasLineOfSight(x1, y1, x2, y2, barriers) {
  for (const barrier of barriers) {
    if (barrier.type === "wall") {
      const locs = barrier.locations;
      if (!locs || locs.length < 2) continue;
      for (let i = 0; i < locs.length - 1; i++) {
        if (segmentsIntersect(x1, y1, x2, y2, locs[i].x, locs[i].y, locs[i + 1].x, locs[i + 1].y)) return false;
      }
    } else if (barrier.type === "circle") {
      if (!barrier.location) continue;
      if (segmentIntersectsCircle(x1, y1, x2, y2, barrier.location.x, barrier.location.y, barrier.radius ?? 0)) return false;
    }
  }
  return true;
}

// Apply all barrier collisions for a unit at (px, py) with collision radius r.
// barriers is the array from the zone config for the unit's current map.
export function resolveBarrierCollisions(px, py, r, barriers) {
  for (const barrier of barriers) {
    if (barrier.type === "wall") {
      const locs = barrier.locations;
      if (!locs || locs.length < 2) continue;
      for (let i = 0; i < locs.length - 1; i++) {
        [px, py] = pushOutOfSegment(
          px, py, r + WALL_HALF_THICKNESS,
          locs[i].x, locs[i].y,
          locs[i + 1].x, locs[i + 1].y
        );
      }
    } else if (barrier.type === "circle") {
      if (!barrier.location) continue;
      [px, py] = pushOutOfCircle(px, py, r, barrier.location.x, barrier.location.y, barrier.radius ?? 0);
    }
  }
  return [px, py];
}

File: game-server/internal/instance/collision.go
package instance

import (
	"math"

	"github.com/delve-mmo/game-server/internal/instanceconfig"
	"github.com/delve-mmo/game-server/internal/instancestate"
)

const wallHalfThickness = 0.2 // feet; half the rendered wall thickness

// resolveCollisions pushes all units with Radius > 0 out of any barriers on
// their current map. Called after applyMovement each tick.
func resolveCollisions(state *instancestate.InstanceState, zone instanceconfig.Zone) {
	barriersByMap := make(map[string][]instanceconfig.Barrier, len(zone.Maps))
	dimsByMap := make(map[string]instanceconfig.Dimensions, len(zone.Maps))
	for _, m := range zone.Maps {
		barriersByMap[m.Identifier] = m.Barriers
		dimsByMap[m.Identifier] = m.FeetDimensions
	}

	for _, unit := range state.Units {
		if unit.Radius == 0 {
			continue
		}
		x, y := unit.Position.X, unit.Position.Y
		for _, b := range barriersByMap[unit.MapIdentifier] {
			switch b.Type {
			case "wall":
				locs := b.Locations
				for i := 0; i < len(locs)-1; i++ {
					x, y = pushOutOfSegment(x, y, unit.Radius+wallHalfThickness,
						locs[i].X, locs[i].Y, locs[i+1].X, locs[i+1].Y)
				}
			case "circle":
				if b.Location == nil {
					continue
				}
				x, y = pushOutOfCircle(x, y, unit.Radius, b.Location.X, b.Location.Y, b.Radius)
			}
		}
		if d := dimsByMap[unit.MapIdentifier]; d.Width > 0 {
			x = math.Max(0, math.Min(d.Width, x))
			y = math.Max(0, math.Min(d.Height, y))
		}
		unit.Position.X, unit.Position.Y = x, y
	}
}

// restoreUnitsThatCrossedBarriers is a last-resort safety net, run at the
// very end of the tick (after every other step that can move a unit,
// including resolveCollisions itself): for each unit that moved this tick on
// the same map it started the tick on, if the straight line from its
// start-of-tick position to its current position crossed a barrier, the
// unit is snapped back to where it started. This catches cases
// resolveCollisions can't - e.g. applyNPCSeparation shoving a unit clean
// through a wall in one tick, landing it on the far side with no overlap at
// its final position for resolveCollisions to push back out of. A unit
// whose map changed this tick (applyMapTransitions) is left alone; that's a
// legitimate teleport, not a shove-through-wall.
func restoreUnitsThatCrossedBarriers(state *instancestate.InstanceState, prevState *instancestate.InstanceState, zone instanceconfig.Zone) {
	for id, unit := range state.Units {
		prev, ok := prevState.Units[id]
		if !ok || prev.MapIdentifier != unit.MapIdentifier {
			continue
		}
		if prev.Position.X == unit.Position.X && prev.Position.Y == unit.Position.Y {
			continue
		}
		if !instanceconfig.LineOfSightClear(zone, unit.MapIdentifier, prev.Position.X, prev.Position.Y, unit.Position.X, unit.Position.Y) {
			unit.Position.X, unit.Position.Y = prev.Position.X, prev.Position.Y
		}
	}
}

// pushOutOfSegment returns (px, py) pushed outside radius r of the closest
// point on segment (ax,ay)→(bx,by). Returns (px,py) unchanged if no overlap.
func pushOutOfSegment(px, py, r, ax, ay, bx, by float64) (float64, float64) {
	dx, dy := bx-ax, by-ay
	lenSq := dx*dx + dy*dy
	if lenSq == 0 {
		return px, py
	}
	t := ((px-ax)*dx + (py-ay)*dy) / lenSq
	if t < 0 {
		t = 0
	} else if t > 1 {
		t = 1
	}
	cx, cy := ax+t*dx, ay+t*dy
	ex, ey := px-cx, py-cy
	dist := math.Sqrt(ex*ex + ey*ey)
	if dist >= r {
		return px, py
	}
	if dist == 0 {
		// Degenerate: push perpendicular to segment direction
		segLen := math.Sqrt(lenSq)
		return px + (-dy/segLen)*r, py + (dx/segLen)*r
	}
	overlap := r - dist
	return px + (ex/dist)*overlap, py + (ey/dist)*overlap
}

// pushOutOfCircle returns (px, py) pushed outside (unitRadius + barrierRadius)
// of the barrier center (cx, cy). Returns (px, py) unchanged if no overlap.
func pushOutOfCircle(px, py, unitRadius, cx, cy, barrierRadius float64) (float64, float64) {
	dx, dy := px-cx, py-cy
	dist := math.Sqrt(dx*dx + dy*dy)
	minDist := unitRadius + barrierRadius
	if dist >= minDist {
		return px, py
	}
	if dist == 0 {
		return px, py + minDist
	}
	overlap := minDist - dist
	return px + (dx/dist)*overlap, py + (dy/dist)*overlap
}
